### ISSUE TITLE
Plugin request for Buf Schema Registry - prost plugins

### ISSUE THREAD

**Issue by mfridman** [bufbuild/plugins#96]
Do we want any of these?

https://buf.build/prost/plugins

- prost
- serde
- tonic
- crate

**Comment by neoeinstein** [comment]
As the author of these plugins, let me know if there's anything I can do to help, or that would contribute to the goals here.

**Comment by mfridman** [comment]
Hey @neoeinstein, thanks for dropping a comment. 

Any chance you could link us to the source code for these plugin(s)?

Also, do these plugins operate like standard Protobuf plugins that take a `CodeGeneratorRequest` on stdin and return a `CodeGeneratorResponse` on stdout? Apologies, it'd been a while since I've looked at these..

Part of open-sourcing this repository and packaging these plugins is for transparency and exposing more information to end-users about the plugins, their documentation, source, etc.

**Comment by neoeinstein** [comment]
Indeed, they do. The development for these plugins is at [neoeinstein/protoc-gen-prost](https://github.com/neoeinstein/protoc-gen-prost)

I haven't set up a _full_ docker build for these just yet, but it shouldn't be to difficult for me to make that happen. Generally, I have been using `cross` to cross-compile the source for musl and then adding the compiled binaries into a `scratch` docker image, but I could look at setting up a Dockerfile that compiles each of the plugins completely inside a Dockerfile.

**Comment by neoeinstein** [comment]
Not the fastest build in the world, but I have a `Dockerfile` prepared that can be run serially to produce all four of the plugins. The first run will take the lion's share of time, but each plugin after that should be relatively fast (since the workspace can share the same base docker images). On my host, it was 7m, 49s, 45s, and 1m44s respectively from nothing. See the linked PR for the updated `Dockerfile` and `publish_buf.sh` scripts.

**Comment by mfridman** [comment]
Awesome, I'll try to take a look into this as soon as I get some cycles.

**Comment by mfridman** [comment]
Hey @neoeinstein, what are your thoughts wrt the versioning question? One requirement we currently have for managed plugins is valid semver.

I took a brief look and it seems like you're publishing to crates.io? So technically we can fetch and package the plugins from there, right?

https://crates.io/search?q=protoc-gen-prost

**Comment by neoeinstein** [comment]
Indeed we _do_ use semantic versioning. If you're talking about the versioning published to BSR, we were following the [best practice that was being mentioned from Buf itself](https://docs.buf.build/bsr/remote-generation/plugin-example#4-build-the-dockerfile), which is to include an additional `-#` in the event of republishing based off of a packaging change, but that didn't include a change to the underlying code.

You can pull the crate directly from crates.io if desired. You'll still need to compile it, as the crate repository is just source code, and not binaries. The `Dockerfile` would be as follows:

```Dockerfile
FROM rust:1.65.0-alpine3.16@sha256:9aa6c9c3df9cfa2f8c5a648a8f1baf8fac31fc5fc7e916389296df92daccf388 as builder
RUN apk add --no-cache musl-dev
WORKDIR /app
ARG BIN
ARG BIN_VERSION
RUN cargo install $BIN --version $BIN_VERSION --locked --root /app

FROM scratch
ARG BIN
COPY --from=builder /app/bin/$BIN /plugin
ENTRYPOINT ["/plugin"]
```

Where `BIN` and `BIN_VERSION` are currently:
* `protoc-gen-prost` `0.2.1`
* `protoc-gen-prost-serde` `0.2.1`
* `protoc-gen-prost-crate` `0.3.1`
* `protoc-gen-tonic` `0.2.1`

**Comment by mfridman** [comment]
Thanks for that info @neoeinstein. 

An update on this issue, we need to add support fetching semver from an upstream source. This should be unblocked by #252.

I hope we can get this metadata from either the crates.io index, or, even better, get approval to "crawl" the API directly.

**Comment by mfridman** [comment]
Hey @LuisOsta, do you find the `protoc-gen-prost-crate` plugin useful as a **remote plugin** even though it doesn't work like the local equivalent?

Do you use this plugin with an option, like `no_features` ?

**Comment by LuisOsta** [comment]
@mfridman Yes it's been great as a remote plugin! You can see exactly how we use it here in one of our open source repositories - https://github.com/knox-networks/ssi/blob/main/registry_resolver/buf.gen.yaml. It's obviously less convenient compared to the remote packages Buf has with Golang. But it's still been really great

**Comment by mfridman** [comment]
That example is using `protoc-gen-prost` and `protoc-gen-tonic`. I can see those working, but what about the `protoc-gen-prost-crate` plugin?

This one: https://crates.io/crates/protoc-gen-prost-crate

Since **remote plugins** on the BSR don't have the file context, that plugin doesn't work and requires some options? So I'm wondering if this is even useful, and those options such as `no_features` is used?

**Comment by LuisOsta** [comment]
Oh my apologies, some wires must have gotten crossed. The specific plugin you mentioned `protoc-gen-prost-crate` couldn't really work for us at all. I think I even tried `no_features` but ended up just deciding to not use it at all. It could be that I simply missed some way to make it work but in the end I decided that using `protoc-gen-prost` and `protoc-gen-tonic` to essentially generate "source" files was the best way to go.

**Comment by neoeinstein** [comment]
Happy to add a LICENSE file to the root if that will help. `crates.io` pulls the license identifier from the `Cargo.toml` specification, but being explicit with a license file in the repository is not a problem.

**Comment by mfridman** [comment]
The `protoc-gen-prost` plugin is available at https://buf.build/community/neoeinstein-prost, added in https://github.com/bufbuild/plugins/pull/288

Will tackle [protoc-gen-prost-serde](https://crates.io/crates/protoc-gen-prost-serde) and [protoc-gen-tonic](https://crates.io/crates/protoc-gen-tonic) in a separate PR.

`protoc-gen-prost-crate` is out of scope for now.
